api/src/wfl/module/covid.clj

```diff
@@ -180,25 +180,135 @@
 
 (defn ^:private create-tdr-source [tx id request]
   (let [create  "CREATE TABLE %s OF TerraDataRepoSourceDetails (PRIMARY KEY (id))"
+        alter   "ALTER TABLE %s ALTER COLUMN id ADD GENERATED ALWAYS AS IDENTITY"
         details (format "TerraDataRepoSourceDetails_%09d" id)
-        _       (jdbc/execute! tx [(format create details)])
+        _       (jdbc/db-do-commands tx [(format create details)
+                                         (format alter details)])
         items   (-> (select-keys request (keys tdr-source-serialized-fields))
                     (set/rename-keys tdr-source-serialized-fields)
                     (assoc :details details)
                     (->> (jdbc/insert! tx tdr-source-table)))]
     [tdr-source-type (-> items first :id str)]))
 
+(defn ^:private find-new-rows
+  "Find new rows in TDR by querying between `last_checked` and the
+   frozen `now`."
+  [{:keys [dataset
+           dataset_table
+           table_column_name
+           last_checked] :as _source}
+   now]
+  (-> (datarepo/query-table-between
+       dataset
+       dataset_table
+       table_column_name
+       [last_checked now]
+       [:datarepo_row_id])
+      :rows
+      flatten))
+
+(defn ^:private go-create-snapshot!
+  "Create snapshot in TDR from `dataset` body, `table`
+   and `row-ids` then write job info as well as rows
+   into `source-details-name` table, `suffix` will be
+   appended to the snapshot names."
+  [suffix dataset table row-ids]
+  (let [columns     (-> (datarepo/all-columns dataset table)
+                        (->> (map :name) set)
+                        (conj "datarepo_row_id"))
+        job-id (-> (datarepo/make-snapshot-request dataset columns table row-ids)
+                   (update :name #(str % suffix))
+                   (datarepo/create-snapshot-job))]
+    job-id))
+
+(defn ^:private create-snapshots
+  "Create uniquely named snapshots in TDR with max partition size of 500,
+   using the frozen `now-obj`, from `row-ids`, return shards and TDR job-ids."
+  [{:keys [dataset dataset_table] :as _source} now-obj row-ids]
+  (let [shards (partition-all 500 row-ids)
+        compact-now (.format now-obj (DateTimeFormatter/ofPattern "YYYYMMdd'T'HHmmss"))
+        job-ids (doall (map-indexed (fn [idx shard]
+                                      (go-create-snapshot!
+                                       (format "_%s_%s" compact-now idx)
+                                       dataset
+                                       dataset_table
+                                       shard)) shards))]
+    [shards job-ids]))
```

Comment:
```clojure
(let [dt-format   (DateTimeFormatter/ofPattern "YYYYMMdd'T'HHmmss")
      compact-now (.format now-obj dt-format)]
  (letfn [(create-snapshot [idx shard]
            [shard (go-create-snapshot! 
                     (format "_%s_%s" compact-now idx)
                      dataset
                      dataset_table
                      shard)])]
    (->> row-ids
        (partition-all 500)
        (map-indexed create-snapshot)
        vec)))
```

Reply:
I think I will keep the `create-snapshot` a separate function for now, for the convenience of testing, if you don't mind.